### DISCUSSION TITLE
`unsafe` policy

### DISCUSSION THREAD

**Discussion by Kixunil** [rust-bitcoin/rust-bitcoin#1228] [General]
There are distinct `unsafe` policies in different crates, some seem very lax (`secp256k1`), other very restrictive (`forbid(unsafe_code)` in `bitcoin`). I believe `unsafe` is not evil but can be dangerous. I wasn't sure where to draw the line but I think it became clearer to me recently.

`unsafe` *can* be written correctly but there were instances of people messing it up. From what I've seen they were mostly "nobody looked at it hard enough for long enough" cases. We can conclude that the presence of `unsafe` significantly increases maintenance cost of the affected chunk of code. The affected chunk of code is very often the whole module containing `unsafe` block or `unsafe` trait/impl. In the remaining cases `unsafe` affects only the function it's used in. In rare cases `unsafe` implicitly spills outside of module which is widely considered wrong design and leads to flame wars in the Rust community.

The maintenance costs include:

* review by multiple independent people highly skilled in `unsafe` Rust - minority of the developers
* doing super-deep review every time the affected code is touched
* running additional analysis tool such as miri, sanitizers, fuzzers, linters
* re-reviewing `unsafe` whenever new information about Rust `unsafe` is published (updates to memory model, things like `MaybeUninit`...)

So naturally the decision should be "use `unsafe` whenever the benefits outweigh the maintenance cost". This is still subjective but hopefully more tangible than other ways of looking at it. To help think about it I identified these categories of `unsafe`:

* FFI calls to libraries with C interface
* API-improving - things like unsized types that can not be expressed without `unsafe` but improve the API
* Basic data structures - `Vec`, `Mutex`...
* Perf improvements
* Fake perf improvements

My preferences/suggestions:

To justify FFI calls the library in question has to be sufficiently large or security-critical to not want to rewrite it in Rust. `libsecp256k1` is a good example. Small, easy things should be just rewritten.
API-improving `unsafe` should be justifiable for things that are used often. Although, it may be better to accept `unsafe` for less frequently used things so that API doesn't need to be broken in the future.
Our crates don't implement basic data structures, so that should be out of scope.
Perf improvements should not be premature. Either the improvement must be very large or someone must explicitly need it and demonstrate it by providing additional support - reviewing the code, writing tests, integrating tools or donating sats to the maintainers.

Fake perf improvements are a special category where someone thinks there's a perf improvement but it's trivially possible to achieve the same using safe code taking into account compiler optimizations. As an example consider this code:

```rust
if slice.len() < 2 {
    return Err("slice too short");
}
// we just checked the length
let first = unsafe { slice.get_unchecked(1) };
let second = unsafe { slice.get_unchecked(2) };
```

This is considered fake optimization because the compiler would remove bound check anyway. It's commonly caused by authors lack of knowledge and is missing benches to prove it helps anything. While this may seem to be totally unacceptable I could imagine it being useful when debug mode is required to debug some other issue and the code would be too slow. However it is better to temporarily patch the code while debugging and keeping it safe under normal circumstances.

So as a co-maintainer I will happily review API-improving `unsafe`, performance optimizations that are easily understandable, justified by benches and extra work by whoever needs them, and  somewhat less happily but out of necessity FFI code of large libraries. I believe I have a decent understanding of Rusts `unsafe` since in the past I've found a few subtle soundness issues in some (even popular) Rust crates.

**Comment by apoelstra** [comment]
Agreed on all counts.

**Comment by dpc** [comment]
IMO `unsafe` must be justified by at least one of:

* there's no way to avoid using it (like C bindings)
* it was _demonstrated_ that using it leads to _significant_ performance improvements _in performance sensitive code_

**Reply by Kixunil** [reply]
That sounds very similar to what I say. If I understand it correctly the difference is:

* You require it's impossible to avoid it (stronger requirement), I require that it's very costly to avoid it. However this is literal interpretation and I guess maybe you didn't actually mean *impossible* because that would imply you require `libsecp256k1` to be rewritten in Rust which I don't expect you intended?
* You require that *significant* performance improvement is *demonstrated* in *performance sensitive* code but don't define what it is and don't require anything more. I say that demonstration is by benchmarks and that "significant" and "performance sensitive" is defined by "someone is willing to spend significant resources helping maintaining it" (thus if they consider it sensitive enough they will pay the cost, if they don't care there will be no `unsafe`)

Do you disagree with anything specific I said?

**Reply by dpc** [reply]
:100:

**Comment by tcharding** [comment]
I guess now is a good time to read https://doc.rust-lang.org/nomicon/ then?

**Reply by Kixunil** [reply]
Just when you review unsafe code. (LOL that intro)
